gui/webdisplay/src/TWebWindow.cxx

```diff
@@ -220,6 +365,109 @@ void ROOT::Experimental::TWebWindow::InvokeCallbacks(bool force)
+/// Key is random number generated when starting new window
+/// procid is special information about starting process which can be used later to halt it
+
+unsigned ROOT::Experimental::TWebWindow::AddProcId(bool batch_mode, const std::string &key, const std::string &procid)
+{
+   std::lock_guard<std::mutex> grd(fConnMutex);
+
+   ++fConnCnt;
+
+   fKeys.emplace_back(std::make_shared<WebConn>(fConnCnt, batch_mode, key, procid));
+
+   return fConnCnt;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Returns true if provided key value already exists (in processes map or in existing connections)
+
+bool ROOT::Experimental::TWebWindow::HasKey(const std::string &key)
+{
+   std::lock_guard<std::mutex> grd(fConnMutex);
+
+   for (auto &&entry : fKeys) {
+      if (entry->fKey == key)
+         return true;
+   }
+
+   for (auto &&conn : fConn) {
+      if (conn->fKey == key)
+         return true;
+   }
+
+   return false;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Check if started process(es) establish connection. After timeout such processed will be killed
+/// Method invoked from http server thread, therefore appropriate mutex must be used on all relevant data
+
+void ROOT::Experimental::TWebWindow::CheckWebKeys()
+{
+   if (!fMgr) return;
+
+   auto stamp = std::chrono::system_clock::now();
+
+   float tmout = fMgr->GetLaunchTmout();
+
+   std::vector<std::string> procs;
+
+   {
+      std::lock_guard<std::mutex> grd(fConnMutex);
+
+      for (auto n = fKeys.size(); n > 0; --n) {
+         std::chrono::duration<double> diff = stamp - fKeys[n - 1]->fSendStamp;
+         // introduce large timeout
+         if (diff.count() > tmout) {
+            R__DEBUG_HERE("webgui") << "Halt process " <<  fKeys[n - 1]->fProcId << " after " << diff.count() << " sec";
+            procs.emplace_back(fKeys[n - 1]->fProcId);
+            fKeys.erase(fKeys.begin() + n - 1);
+         }
+      }
+   }
+
+   for (auto &&entry : procs)
+      fMgr->HaltClient(entry);
+}
+
+
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Check if there are connection which are inactive for longer time
+/// For instance, batch browser will be stopped if no activity for 30 sec is there
+
+void ROOT::Experimental::TWebWindow::CheckInactiveConnections()
+{
+   auto stamp = std::chrono::system_clock::now();
+
+   double batch_tmout = 20.;
+
+   std::vector<std::shared_ptr<WebConn>> clr;
+
+   {
+      std::lock_guard<std::mutex> grd(fConnMutex);
+
+      for (auto n = fConn.size(); n > 0; --n) {
+         std::chrono::duration<double> diff = stamp - fConn[n-1]->fSendStamp;
+         // introduce large timeout
+         if ((diff.count() > batch_tmout) && fConn[n-1]->fBatchMode) {
+            fConn[n-1]->fActive = false;
+            clr.push_back(std::move(fConn[n-1]));
+            fConn.erase(fConn.begin() + n - 1);
```

Comment:
See iter/remove pattern above

Reply:
Yes, but here same problem - why to touch existing vector if probability to change something is less then 0.0001

Reply:
Please apply same erase/remove pattern from above.

Reply:
> Please apply same erase/remove pattern from above.

I explained already - in very rare cases (less 1/10000) something will be changed in connections. And I do not want to permanently do std::move(), std::swap() operations.

Reply:
I am talking about the approach from https://github.com/root-project/root/pull/2603/files/f4cdac1cf45a42f3629251adc48976bf8f0a80b5#r218733309 which has no "nothing to do" branch, but get rids of the fragile approach of "iterate over indices and erase elements during iteration"

Reply:
Sorry, now I see. I add code with std::remove_if()